### DISCUSSION TITLE
Open url from text file in browser when there is text in the same line of the url

### DISCUSSION THREAD

**Discussion by ghost** [junegunn/fzf#3341] [Q&A]
I am using this to open urls from a file in the browser

`grep -n '' "filename" | fzf -m --reverse | xargs -d '\n' firefox`

It works well when I only have the url in a line. The problem is that in that file I have the url with more text.  I use it for bookmarks and it looks like this per line: 

`"Bookmark Name" | https://github.com/junegunn/fzf/ | +tag1 +tag2 | [Description-optional]`

What I want is to open only the url and not the entire line in the browser. I have searched before posting this, so help would be appreciated.

**Comment by hamvocke** [comment]
I've written [a small script](https://github.com/hamvocke/lnks) that tries to do something very similar to what you're describing here. Maybe this is helpful.

My solution uses fzf's *"Field Index Expressions"* to pull this off (see `man fzf` or check it out [here](https://github.com/junegunn/fzf/blob/76364ea767cca7ca8f6570a049fbb8d3fff751a9/man/man1/fzf.1#L842)). It might not be quite what you're looking for given you're passing the `--multi` flag.

To solve your particular problem, you could try to sanitize `fzf`'s output with `cut` before passing it over to `firefox`. Something like this might work (assuming the URLs are always in the second field of each line):

```bash
cat yourfile.txt | fzf --multi --reverse | cut -d '|' -f 2 | xargs firefox
```

**Reply by LangLangBart** [reply]
- with the `--delimiter` flag it can be easier, it defines how the fields are separated, in this case an escaped `|` (pipe) character
  - the default delimiter seems to be defined like this `\S+\s*`

https://github.com/junegunn/fzf/blob/76364ea767cca7ca8f6570a049fbb8d3fff751a9/src/tokenizer.go#L147

- the `{+}` is an additional placeholder flag to get all selected lines
- `become` action was introduced in [0.38.0](https://github.com/junegunn/fzf/blob/master/CHANGELOG.md#0380)
-  open a url in your default browser with `open`, the command may differ on `Linux/Windows`

```sh
cat yourfile.txt | fzf --multi --reverse --delimiter '\|'  --bind 'enter:become:open {+2}'
```

**Comment by ghost** [comment]
Thanks @hamvocke and @LangLangBart. Both of the solutions work but I also need this: 

- Make ` - ` the delimiter instead of  ` | ` because I am using that delimiter right now. The problem is that some urls also have `-` and changing to --delimiter `\ - `  does not work.
-  Maintain opened the terminal after opening any url. Currently, with those commands, after opening one url the terminal exits fzf. I would like to continue with it open to search more urls instead of having to reopen it.
- Open the url no matter the exact position. There are some of my urls that are not in the second position, so none of those commands work in this case.

Apart from this, I do not know if there is any shortcut to select all the input in the search bar of fzf (like ctrl+a). Any idea?

**Reply by LangLangBart** [reply]
>  Open the url no matter the exact position

You can use grep to search for a typical URL with a regex and pass the found URLs to `fzf`. Disadvantage: You lose the context of some bookmarks.

```bash
grep -Eo 'https?://[^ ]*' <YOUR FILE> | fzf --multi --reverse --bind 'ctrl-a:toggle-all' --bind 'enter:execute-silent:open {+}'
```

**Reply by LangLangBart** [reply]
- this could be an alternative using the `{+f}` placeholder
   - see `man fzf` for details 

```bash
grep -n '' <YOUR FILE> | fzf --multi --reverse --bind 'ctrl-a:toggle-all' --bind "enter:execute-silent:grep -Eo 'https?://[^ ]*' {+f} | xargs open"
```

**Reply by ghost** [reply]
The first command works but, as you say, I lose a ton of context. 

The second one does no open anything. When I press ENTER nothing happens (no url is opened).

**Reply by LangLangBart** [reply]
- redirecting the errors to  `/tmp/err` might shed some light, why the second command won't work 

```bash
grep -n '' <YOUR FILE> | fzf --multi --reverse --bind 'ctrl-a:toggle-all' --bind "enter:execute-silent:(grep -Eo 'https?://[^ ]*' {+f} | xargs open) 2>/tmp/err"

cat /tmp/err
```

**Reply by ghost** [reply]
I have used that command and then `cat /tmp/err` but nothing is created (I receive this message: cat: /tmp/err: No such file or directory).
